platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileCompactor.java

```diff
@@ -0,0 +1,479 @@
+            return;
+        }
+        final int filesCount = filesToCompact.size();
+        if (filesCount < config.minNumberOfFilesInMerge()) {
+            logger.debug(
+                    MERKLE_DB.getMarker(),
+                    "[{}] No need to merge as {} is less than the minimum {} files to merge.",
+                    storeName,
+                    filesCount,
+                    config.minNumberOfFilesInMerge());
+            return;
+        }
+
+        final long start = System.currentTimeMillis();
+
+        final long filesToMergeSize = getSizeOfFiles(filesToCompact);
+        logger.debug(
+                MERKLE_DB.getMarker(),
+                "[{}] Starting merging {} files / {}",
+                storeName,
+                filesCount,
+                formatSizeBytes(filesToMergeSize));
+
+        final List<Path> newFilesCreated = compactFiles(index, filesToCompact);
+
+        final long end = System.currentTimeMillis();
+        final long tookMillis = end - start;
+        if (reportDurationMetricFunction != null) {
+            reportDurationMetricFunction.accept(compactionLevel, tookMillis);
+        }
+
+        final long mergedFilesSize = getSizeOfFilesByPath(newFilesCreated);
+        if (reportSavedSpaceMetricFunction != null) {
+            reportSavedSpaceMetricFunction.accept(
+                    compactionLevel, (filesToMergeSize - mergedFilesSize) * UnitConstants.BYTES_TO_MEBIBYTES);
+        }
+
+        logMergeStats(storeName, tookMillis, filesToCompact, filesToMergeSize, newFilesCreated, dataFileCollection);
+        logger.debug(
+                MERKLE_DB.getMarker(),
+                "[{}] Finished merging {} files / {} in {} ms",
+                storeName,
+                filesCount,
+                formatSizeBytes(filesToMergeSize),
+                tookMillis);
+    }
+
+    private boolean isTimeForFullMerge(final Instant startMerge) {
+        return startMerge
+                .minus(config.fullMergePeriod(), config.mergePeriodUnit())
+                .isAfter(lastFullMerge);
+    }
+
+    private boolean isTimeForMediumMerge(final Instant startMerge) {
+        return startMerge
+                .minus(config.mediumMergePeriod(), config.mergePeriodUnit())
+                .isAfter(lastMediumMerge);
+    }
+
+    /**
+     * Create a filter to only return all new files that belong to certain compaction level
+     *
+     * @param expectedCompactionLevel compaction level to filter by
+     * @param maxNumberOfFilesInMerge The maximum number of files to process in a single merge
+     * @return filter to filter list of files
+     */
+    public static UnaryOperator<List<DataFileReader<?>>> readersOfLevel(
+            int expectedCompactionLevel, int maxNumberOfFilesInMerge) {
+        return dataFileReaders -> {
+            final List<DataFileReader<?>> filesNewestFirst = dataFileReaders.stream()
+                    .sorted(DATA_FILE_READER_CREATION_TIME_COMPARATOR_REVERSED)
```

Comment:
Why is sorting by creation time needed here?

Reply:
I assumed that it could be useful for more optimal traversal during compaction. But since you're asking, I'd guess that it's not needed, so I'll remove it.